torch_xla/csrc/ops/device_data.h

```diff
@@ -22,6 +22,13 @@ class DeviceData : public XlaNode {
     return data_;
   }
 
+  // With SPMD sharding propagation, we need to update the unpartitioned
+  // backend data with a partitioned one in the node operands. The node and its
+  // device data operands are fixed, but the underlying device data address need
+  // to be updated. An alternative to `Assign` would be make the operands of
+  // nodes mutable and modify with new device data nodes.
+  void Assign(std::shared_ptr<torch::lazy::BackendData> data) { data_ = data; }
```

Comment:
we should at very least check if shape matches before modifying the underlying data